crates/paralegal-policy/src/flows_to.rs

```diff
@@ -166,69 +166,94 @@ impl CtrlFlowsTo {
+    pub fn data_and_control_flows_to(
+        &self,
+        ctrl: &Ctrl,
+        src: DataSource,
+        sink: CallSiteOrDataSink,
+    ) -> bool {
+        let mut influencees = DataAndControlInfluencees::new(src, ctrl, self);
+        influencees.contains(&sink)
+    }
+}
 
-        while let Some(cur_src) = queue.pop() {
-            if let DataSource::FunctionCall(callsite) = &cur_src {
-                if CallSiteOrDataSink::CallSite(callsite.clone()) == sink {
-                    return true;
-                }
-            }
+pub(crate) struct DataAndControlInfluencees<'a> {
+    to_return: Vec<CallSiteOrDataSink>,
+    queue: Vec<DataSource>,
+    seen: std::collections::HashSet<CallSiteOrDataSink>,
+    ctrl: &'a Ctrl,
+    flows_to: &'a CtrlFlowsTo,
+}
 
-            let cur_src_index = cur_src.clone().to_index(&self.sources);
-            for cur_sink in self
-                .data_flows_to
-                .row_set(&cur_src_index)
-                .iter()
-                .chain(self.ctrl_flows_to.row_set(&cur_src_index).iter())
-            {
-                if &sink == cur_sink {
-                    return true;
-                }
+impl<'a> DataAndControlInfluencees<'a> {
+    pub(crate) fn new(src: DataSource, ctrl: &'a Ctrl, flows_to: &'a CtrlFlowsTo) -> Self {
+        let queue = vec![src];
+        let seen = std::collections::HashSet::<CallSiteOrDataSink>::new();
+
+        DataAndControlInfluencees {
+            to_return: Vec::new(),
+            queue,
+            seen,
+            ctrl,
+            flows_to,
+        }
+    }
+}
+
+impl<'a> Iterator for DataAndControlInfluencees<'a> {
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(r) = self.to_return.pop() {
+            return Some(r);
+        }
+        if let Some(cur_src) = self.queue.pop() {
+            let cur_src_index = cur_src.clone().to_index(&self.flows_to.sources);
+            for cur_sink in self.flows_to.data_flows_to.row_set(&cur_src_index).iter() {
+                self.to_return.push(cur_sink.clone());
 
                 let cur_sink_callsite = match &cur_sink {
                     CallSiteOrDataSink::CallSite(cs) => cs,
                     CallSiteOrDataSink::DataSink(DataSink::Argument { function, .. }) => function,
                     _ => continue,
                 };
-                if seen.insert(cur_sink) {
-                    queue.push(cur_sink_callsite.clone().into());
+                if self.seen.insert(cur_sink.clone()) {
+                    self.queue.push(cur_sink_callsite.clone().into());
                 }
             }
-        }
 
-        false
+            if let Some(callsites) = self.ctrl.ctrl_flow.get(&cur_src) {
+                for cur_cs_sink in callsites {
+                    let cs_or_ds: CallSiteOrDataSink = cur_cs_sink.clone().into();
+                    self.to_return.push(cs_or_ds.clone());
+                    for arg in self
```

Comment:
I think this could be simplified to 

```rs
self.to_return.extend(
    self.flows_to
        .callsites_to_callargs
        .row(&cs_or_ds.clone().to_index(&self.flows_to.sinks))
        .cloned()
)
```